File: tools/python/reroom_verify/coordinate.py
"""Pure RR-COORD-1 coordinate, projection, orientation, and float helpers."""

from __future__ import annotations

import json
import math
import struct
from typing import Any, Sequence

from .canonical_json import artifact
from .loader import LoadedFixture, VerificationFailure, parse_json_bytes


FLOAT32_MAX = 3.4028234663852886e38
SCALAR_ABSOLUTE_TOLERANCE = 1e-5
SCALAR_RELATIVE_TOLERANCE = 1e-6
RIGID_TOLERANCE = 1e-4
HOMOGENEOUS_ROW_TOLERANCE = 1e-6


def _number(value: Any) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise VerificationFailure("coordinate_invalid", "coordinate scalar is not numeric")
    converted = float(value)
    if not math.isfinite(converted):
        raise VerificationFailure("numeric_out_of_range", "coordinate scalar is non-finite")
    if abs(converted) > FLOAT32_MAX:
        raise VerificationFailure("numeric_out_of_range", "coordinate scalar exceeds binary32")
    try:
        quantized = struct.unpack(">f", struct.pack(">f", converted))[0]
    except OverflowError as error:
        raise VerificationFailure("numeric_out_of_range", "coordinate scalar exceeds binary32") from error
    if not math.isfinite(quantized):
        raise VerificationFailure("numeric_out_of_range", "coordinate scalar overflows binary32")
    return quantized


def _vector(values: Any, length: int) -> list[float]:
    if not isinstance(values, list) or len(values) != length:
        raise VerificationFailure("coordinate_invalid", "coordinate vector has wrong shape")
    return [_number(value) for value in values]


def _matrix(values: Any, dimension: int) -> list[list[float]]:
    flat = _vector(values, dimension * dimension)
    return [flat[row * dimension : (row + 1) * dimension] for row in range(dimension)]


def multiply_matrix_vector(matrix: Sequence[Sequence[float]], vector: Sequence[float]) -> list[float]:
    if not matrix or any(len(row) != len(vector) for row in matrix):
        raise VerificationFailure("coordinate_invalid", "matrix/vector shapes disagree")
    return [sum(coefficient * value for coefficient, value in zip(row, vector)) for row in matrix]


def multiply_matrices(left: Sequence[Sequence[float]], right: Sequence[Sequence[float]]) -> list[list[float]]:
    if not left or not right or any(len(row) != len(right) for row in left):
        raise VerificationFailure("coordinate_invalid", "matrix shapes disagree")
    columns = list(zip(*right))
    return [
        [sum(a * b for a, b in zip(row, column)) for column in columns]
        for row in left
    ]


def _determinant3(matrix: Sequence[Sequence[float]]) -> float:
    return (
        matrix[0][0] * (matrix[1][1] * matrix[2][2] - matrix[1][2] * matrix[2][1])
        - matrix[0][1] * (matrix[1][0] * matrix[2][2] - matrix[1][2] * matrix[2][0])
        + matrix[0][2] * (matrix[1][0] * matrix[2][1] - matrix[1][1] * matrix[2][0])
    )


def validate_rigid_transform(values: Any) -> list[list[float]]:
    matrix = _matrix(values, 4)
    if any(abs(matrix[3][index] - expected) > HOMOGENEOUS_ROW_TOLERANCE for index, expected in enumerate((0, 0, 0, 1))):
        raise VerificationFailure("coordinate_invalid", "homogeneous row is invalid")
    rotation = [row[:3] for row in matrix[:3]]
    orthogonality_squared = 0.0
    for row in range(3):
        for column in range(3):
            dot = sum(rotation[index][row] * rotation[index][column] for index in range(3))
            difference = dot - (1.0 if row == column else 0.0)
            orthogonality_squared += difference * difference
    if math.sqrt(orthogonality_squared) > RIGID_TOLERANCE:
        raise VerificationFailure("coordinate_invalid", "rotation is not orthonormal")
    if abs(_determinant3(rotation) - 1.0) > RIGID_TOLERANCE:
        raise VerificationFailure("coordinate_invalid", "rotation must be proper, not reflected")
    return matrix


def rr_float_equal(left: Any, right: Any) -> bool:
    a = _number(left)
    b = _number(right)
    return abs(a - b) <= SCALAR_ABSOLUTE_TOLERANCE + SCALAR_RELATIVE_TOLERANCE * max(abs(a), abs(b))


def _validate_rigid_matrix3(values: Any) -> list[list[float]]:
    matrix = _matrix(values, 3)
    orthogonality_squared = 0.0
    for row in range(3):
        for column in range(3):
            dot = sum(matrix[index][row] * matrix[index][column] for index in range(3))
            difference = dot - (1.0 if row == column else 0.0)
            orthogonality_squared += difference * difference
    if (
        math.sqrt(orthogonality_squared) > RIGID_TOLERANCE
        or abs(_determinant3(matrix) - 1.0) > RIGID_TOLERANCE
    ):
        raise VerificationFailure("coordinate_invalid", "camera conversion is not a proper rotation")
    return matrix


def apply_world_correction(document: dict[str, Any]) -> dict[str, Any]:
    source = document.get("from_world_frame_version")
    target = document.get("to_world_frame_version")
    if (
        isinstance(source, bool)
        or isinstance(target, bool)
        or not isinstance(source, int)
        or not isinstance(target, int)
        or source < 1
        or target <= source
    ):
        raise VerificationFailure("coordinate_invalid", "correction must move strictly forward")
    matrix = validate_rigid_transform(document.get("to_from_from_transform"))
    point = _vector(document.get("point_from"), 4)
    return {"point_to": multiply_matrix_vector(matrix, point)}


def transform_intrinsics(document: dict[str, Any]) -> dict[str, Any]:
    intrinsics = document.get("sensor_intrinsics")
    if not isinstance(intrinsics, dict):
        raise VerificationFailure("coordinate_invalid", "sensor intrinsics are missing")
    fx, fy, cx, cy = (_number(intrinsics.get(key)) for key in ("fx", "fy", "cx", "cy"))
    if fx <= 0 or fy <= 0:
        raise VerificationFailure("coordinate_invalid", "focal lengths must be positive")
    encoded_from_sensor = _matrix(document.get("encoded_from_sensor"), 3)
    if any(abs(encoded_from_sensor[2][index] - expected) > HOMOGENEOUS_ROW_TOLERANCE for index, expected in enumerate((0, 0, 1))):
        raise VerificationFailure("coordinate_invalid", "pixel transform is not affine")
    sensor = [[fx, 0.0, cx], [0.0, fy, cy], [0.0, 0.0, 1.0]]
    encoded = multiply_matrices(encoded_from_sensor, sensor)
    encoded_fx = math.hypot(encoded[0][0], encoded[0][1])
    encoded_fy = math.hypot(encoded[1][0], encoded[1][1])
    if encoded_fx <= 0 or encoded_fy <= 0:
        raise VerificationFailure("coordinate_invalid", "encoded focal lengths are degenerate")
    encoded_size = document.get("encoded_size")
    if (
        not isinstance(encoded_size, list)
        or len(encoded_size) != 2
        or any(isinstance(value, bool) or not isinstance(value, int) or value <= 0 for value in encoded_size)
    ):
        raise VerificationFailure("coordinate_invalid", "encoded size is invalid")
    orientation = document.get("orientation")
    if orientation != "up":
        raise VerificationFailure("coordinate_invalid", "encoded orientation must be physically upright")
    return {
        "encoded_intrinsics": {
            "fx": encoded_fx,
            "fy": encoded_fy,
            "cx": encoded[0][2],
            "cy": encoded[1][2],
        },
        "encoded_size": encoded_size,
        "orientation": orientation,
    }


def validate_rr_float(value: Any) -> dict[str, Any]:
    number = _number(value)
    if abs(number) > FLOAT32_MAX:
        raise VerificationFailure("numeric_out_of_range", "value exceeds finite binary32")
    try:
        encoded = struct.pack(">f", number)
    except OverflowError as error:
        raise VerificationFailure("numeric_out_of_range", "value exceeds finite binary32") from error
    if not math.isfinite(struct.unpack(">f", encoded)[0]):
        raise VerificationFailure("numeric_out_of_range", "binary32 result is non-finite")
    return {"binary32_hex": encoded.hex()}


def project(document: dict[str, Any]) -> dict[str, Any]:
    intrinsics = document.get("intrinsics")
    if not isinstance(intrinsics, dict):
        raise VerificationFailure("coordinate_invalid", "projection intrinsics are missing")
    fx, fy, cx, cy = (_number(intrinsics.get(key)) for key in ("fx", "fy", "cx", "cy"))
    if fx <= 0 or fy <= 0:
        raise VerificationFailure("coordinate_invalid", "focal lengths must be positive")
    point = _vector(document.get("camera_point"), 3)
    if document.get("pixel_center") != "half_integer":
        raise VerificationFailure("coordinate_invalid", "pixel center convention is invalid")
    if point[2] == 0:
        raise VerificationFailure("coordinate_invalid", "point lies on the camera plane")
    return {
        "encoded_pixel": [fx * point[0] / point[2] + cx, fy * point[1] / point[2] + cy],
        "visible": point[2] < 0,
    }


def arkit_to_opencv_camera(document: dict[str, Any]) -> dict[str, Any]:
    point = _vector(document.get("camera_point"), 3)
    conversion = _validate_rigid_matrix3(document.get("conversion"))
    return {"opencv_camera_point": multiply_matrix_vector(conversion, point)}


def evaluate_coordinate(document: dict[str, Any]) -> tuple[str, dict[str, Any]]:
    operation = document.get("operation")
    if operation == "apply_world_correction":
        return "matrix", apply_world_correction(document)
    if operation == "transform_intrinsics":
        return "matrix", transform_intrinsics(document)
    if operation == "validate_rr_float":
        return "runner_consensus", validate_rr_float(document.get("value"))
    if operation == "project":
        return "projected_pixels", project(document)
    if operation == "arkit_to_opencv_camera":
        return "matrix", arkit_to_opencv_camera(document)
    if operation == "validate_rigid_transform":
        validate_rigid_transform(document.get("values"))
        return "matrix", {"valid": True}
    raise VerificationFailure("coordinate_invalid", "unknown coordinate operation")


def _normalize_integral_floats(value: Any) -> Any:
    if isinstance(value, float) and value.is_integer():
        return int(value)
    if isinstance(value, dict):
        return {key: _normalize_integral_floats(child) for key, child in value.items()}
    if isinstance(value, list):
        return [_normalize_integral_floats(child) for child in value]
    return value


def execute_coordinate_case(fixture: LoadedFixture, case: dict[str, Any]) -> list[dict[str, Any]]:
    document = parse_json_bytes(
        fixture.read_fixture_file(case["input"]["relative_path"]),
        max_depth=fixture.max_document_depth,
    )
    if not isinstance(document, dict):
        raise VerificationFailure("coordinate_invalid", "coordinate vector must be an object")
    kind, result = evaluate_coordinate(document)
    raw = (
        json.dumps(
            _normalize_integral_floats(result),
            ensure_ascii=False,
            allow_nan=False,
            separators=(",", ":"),
        ).encode("utf-8")
        + b"\n"
    )
    return [artifact(kind, raw)]

File: tools/javascript/src/coordinate.mjs
import {
  RunnerFailure,
  caseResult,
  parseJsonBytesStrict,
  readFixtureFile,
  rejectedCaseResult,
  sha256Hex,
} from "./loader.mjs";

export const RR_FLOAT_MAX = 3.4028234663852886e38;
const SCALAR_ABSOLUTE_TOLERANCE = 1e-5;
const SCALAR_RELATIVE_TOLERANCE = 1e-6;
const RIGID_TOLERANCE = 1e-4;
const HOMOGENEOUS_ROW_TOLERANCE = 1e-6;

function requireCoordinate(condition, message, rejectionClass = "coordinate_invalid") {
  if (!condition) throw new RunnerFailure(rejectionClass, message);
}

export function quantizeRrFloat(value) {
  requireCoordinate(typeof value === "number" && Number.isFinite(value) && Math.abs(value) <= RR_FLOAT_MAX, "value is outside finite binary32 range", "numeric_out_of_range");
  const quantized = Math.fround(value);
  requireCoordinate(Number.isFinite(quantized), "value overflows binary32", "numeric_out_of_range");
  return quantized;
}

export function rrFloatEqual(left, right) {
  const a = quantizeRrFloat(left);
  const b = quantizeRrFloat(right);
  return Math.abs(a - b) <= SCALAR_ABSOLUTE_TOLERANCE + SCALAR_RELATIVE_TOLERANCE * Math.max(Math.abs(a), Math.abs(b));
}

export function rrTranslationEqual(left, right) {
  return rrFloatEqual(left, right) && Math.abs(quantizeRrFloat(left) - quantizeRrFloat(right)) <= 1e-4;
}

export function rrTransformedIntrinsicsEqual(left, right) {
  return rrFloatEqual(left, right) && Math.abs(quantizeRrFloat(left) - quantizeRrFloat(right)) <= 1e-3;
}

function quantizedArray(values, length) {
  requireCoordinate(Array.isArray(values) && values.length === length, `expected ${length} matrix/vector values`);
  return values.map(quantizeRrFloat);
}

function determinant3(matrix) {
  const [a, b, c, d, e, f, g, h, i] = matrix;
  return a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g);
}

function validateProperRotation(rotation) {
  let orthogonalitySquared = 0;
  for (let row = 0; row < 3; row += 1) {
    for (let column = 0; column < 3; column += 1) {
      let dot = 0;
      for (let index = 0; index < 3; index += 1) dot += rotation[index * 3 + row] * rotation[index * 3 + column];
      const difference = dot - (row === column ? 1 : 0);
      orthogonalitySquared += difference * difference;
    }
  }
  requireCoordinate(Math.sqrt(orthogonalitySquared) <= RIGID_TOLERANCE, "rotation is not orthonormal");
  requireCoordinate(Math.abs(determinant3(rotation) - 1) <= RIGID_TOLERANCE, "rotation determinant is not +1");
  return rotation;
}

export function validateRigidTransform(values) {
  const matrix = quantizedArray(values, 16);
  const rotation = [matrix[0], matrix[1], matrix[2], matrix[4], matrix[5], matrix[6], matrix[8], matrix[9], matrix[10]];
  validateProperRotation(rotation);
  const expectedLastRow = [0, 0, 0, 1];
  requireCoordinate(matrix.slice(12).every((value, index) => Math.abs(value - expectedLastRow[index]) <= HOMOGENEOUS_ROW_TOLERANCE), "homogeneous last row is invalid");
  return matrix;
}

function multiplyMatrixVector(matrix, vector, size) {
  const output = new Array(size).fill(0);
  for (let row = 0; row < size; row += 1) {
    for (let column = 0; column < size; column += 1) output[row] += matrix[row * size + column] * vector[column];
  }
  return output.map((value) => Object.is(value, -0) ? 0 : value);
}

export function projectCameraPoint(intrinsics, cameraPoint) {
  const [x, y, z] = quantizedArray(cameraPoint, 3);
  requireCoordinate(z !== 0, "camera point lies on the projection plane");
  const fx = quantizeRrFloat(intrinsics.fx);
  const fy = quantizeRrFloat(intrinsics.fy);
  const cx = quantizeRrFloat(intrinsics.cx);
  const cy = quantizeRrFloat(intrinsics.cy);
  requireCoordinate(fx > 0 && fy > 0, "intrinsic focal lengths must be positive");
  return {
    encoded_pixel: [fx * (x / z) + cx, fy * (y / z) + cy].map((value) => Object.is(value, -0) ? 0 : value),
    visible: z < 0,
  };
}

export function transformIntrinsics(sensorIntrinsics, encodedFromSensor, encodedSize, orientation) {
  requireCoordinate(orientation === "up", "encoded orientation must be physically upright");
  const transform = quantizedArray(encodedFromSensor, 9);
  const [width, height] = encodedSize;
  requireCoordinate(Number.isInteger(width) && width > 0 && Number.isInteger(height) && height > 0, "encoded size is invalid");
  const matrix = [
    quantizeRrFloat(sensorIntrinsics.fx), 0, quantizeRrFloat(sensorIntrinsics.cx),
    0, quantizeRrFloat(sensorIntrinsics.fy), quantizeRrFloat(sensorIntrinsics.cy),
    0, 0, 1,
  ];
  const product = new Array(9).fill(0);
  for (let row = 0; row < 3; row += 1) {
    for (let column = 0; column < 3; column += 1) {
      for (let index = 0; index < 3; index += 1) product[row * 3 + column] += transform[row * 3 + index] * matrix[index * 3 + column];
    }
  }
  requireCoordinate(Math.abs(product[6]) <= HOMOGENEOUS_ROW_TOLERANCE && Math.abs(product[7]) <= HOMOGENEOUS_ROW_TOLERANCE && Math.abs(product[8] - 1) <= HOMOGENEOUS_ROW_TOLERANCE, "encoded pixel transform is not affine");
  return {
    encoded_intrinsics: {
      fx: Math.hypot(product[0], product[1]),
      fy: Math.hypot(product[3], product[4]),
      cx: product[2],
      cy: product[5],
    },
    encoded_size: [width, height],
    orientation: "up",
  };
}

export function applyWorldCorrection(fromVersion, toVersion, transform, pointFrom) {
  requireCoordinate(Number.isInteger(fromVersion) && Number.isInteger(toVersion) && fromVersion >= 1 && toVersion > fromVersion, "world correction must target a strictly newer epoch");
  return multiplyMatrixVector(validateRigidTransform(transform), quantizedArray(pointFrom, 4), 4);
}

function binary32Hex(value) {
  const bytes = Buffer.alloc(4);
  bytes.writeFloatBE(quantizeRrFloat(value));
  return bytes.toString("hex");
}

export function executeCoordinateOperation(input) {
  switch (input.operation) {
    case "project":
      requireCoordinate(input.pixel_center === "half_integer", "pixel-center convention is unsupported");
      return projectCameraPoint(input.intrinsics, input.camera_point);
    case "transform_intrinsics":
      return transformIntrinsics(input.sensor_intrinsics, input.encoded_from_sensor, input.encoded_size, input.orientation);
    case "arkit_to_opencv_camera":
      return { opencv_camera_point: multiplyMatrixVector(validateProperRotation(quantizedArray(input.conversion, 9)), quantizedArray(input.camera_point, 3), 3) };
    case "apply_world_correction":
      return { point_to: applyWorldCorrection(input.from_world_frame_version, input.to_world_frame_version, input.to_from_from_transform, input.point_from) };
    case "validate_rr_float":
      return { binary32_hex: binary32Hex(input.value) };
    case "validate_rigid_transform":
      validateRigidTransform(input.values);
      return { valid: true };
    default:
      throw new RunnerFailure("coordinate_invalid", "unsupported coordinate operation");
  }
}

export async function executeCoordinateCase(fixture, fixtureCase) {
  try {
    const output = executeCoordinateOperation(parseJsonBytesStrict(await readFixtureFile(fixture, fixtureCase.input)));
    const bytes = Buffer.from(`${JSON.stringify(output)}\n`, "utf8");
    const kind = fixtureCase.expected.artifacts[0]?.kind;
    return caseResult(fixtureCase.case_id, "accept", null, kind ? [{ kind, byte_length: bytes.length, sha256: sha256Hex(bytes) }] : []);
  } catch (error) {
    return rejectedCaseResult(fixtureCase, error);
  }
}
